src/Shared/packet_adapter/packet_parser/infra/packet_parser.go

```diff
@@ -1,142 +1,141 @@
+	enums       map[name]domain.Enum
+}
+
+func NewParser(packets []excelAdapter.PacketDTO) PacketParser {
+	return PacketParser{
+		packetTypes: getPacketTypes(packets),
+		enums:       getEnums(packets),
+	}
+}
+
+func getEnums(packets []excelAdapter.PacketDTO) map[name]domain.Enum {
+	enums := make(map[name]domain.Enum, 0)
+	for _, packet := range packets {
+		for _, measurement := range packet.Measurements {
+			if domain.IsEnum(measurement.ValueType) {
+				enums[measurement.Name] = domain.NewEnum(measurement.ValueType)
+			}
+		}
+	}
+
+	return enums
+}
+
+func getPacketTypes(packets []excelAdapter.PacketDTO) map[uint16]packetMeasurements {
+	packetMeasurements := make(map[id]packetMeasurements, len(packets))
+
+	for _, packet := range packets {
+		measurementDataArr := getMeasurementData(packet)
+		id, err := strconv.ParseUint(packet.Description.ID, 10, 16)
+
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		packetMeasurements[uint16(id)] = measurementDataArr
+	}
+
+	return packetMeasurements
+}
+
+func getMeasurementData(packet excelAdapter.PacketDTO) []domain.MeasurementData {
+	measurementDataArr := make([]domain.MeasurementData, len(packet.Measurements))
+	for index, measurement := range packet.Measurements {
+		valueType := measurement.ValueType
+		if domain.IsEnum(valueType) {
+			valueType = "enum"
+		}
+		measurementData := domain.NewMeasurement(measurement.Name, valueType)
+		measurementDataArr[index] = measurementData
+	}
+	return measurementDataArr
+}
+
+func (parser PacketParser) Decode(data []byte) dto.PacketUpdate {
+	dataReader := bytes.NewBuffer(data)
+	id := serde.DecodeID(dataReader)
+	values := parser.decodePacket(parser.packetTypes[id], dataReader)
+
+	return dto.NewPacketUpdate(id, values, data)
+}
+
+func (parser PacketParser) decodePacket(measurements packetMeasurements, bytes io.Reader) map[name]any {
+	values := make(map[name]any, len(measurements))
+	for _, measurementData := range measurements {
+		values[measurementData.Name()] = parser.decodeMeasurement(measurementData, bytes)
+	}
+	return values
+}
+
+func (parser PacketParser) decodeMeasurement(measurement domain.MeasurementData, reader io.Reader) any {
+	switch measurement.ValueType() {
+	case "enum":
+		return serde.DecodeEnum(parser.enums[measurement.Name()], reader)
+	case "bool":
+		return serde.DecodeBool(reader)
+	case "string":
+		return serde.DecodeString(reader)
+	default:
+		return serde.DecodeNumber(measurement.ValueType(), reader)
+	}
+}
+
+func (parser PacketParser) Encode(packet dto.PacketValues) []byte {
+	dataWriter := bytes.NewBuffer(make([]byte, 0))
+	serde.EncodeID(packet.ID(), dataWriter)
+	for _, measurement := range parser.packetTypes[packet.ID()] {
+		parser.encodeValue(packet.ID(), measurement.Name(), packet.GetValue(measurement.Name()), dataWriter)
+	}
+
+	return dataWriter.Bytes()
+}
+
+func (parser PacketParser) encodeValue(id uint16, name string, value string, bytes io.Writer) {
+	switch parser.findMeasurement(id, name).ValueType() {

```

Comment:
esto es similar a lo de antes, si en vez de 3 casos tuvieras 50 pondrías 50 cases? por otro lado, si dependiendo del caso empiezas a necesitar lógica, empezarías a implementar lógica de negocio distinta todo en el mismo método (y en infra además)? 

estas cosas parecen tonterias, pero cuando llegas a la ehw y hace 4 meses que no tocas este método y de repente peta algo de por aqui, ánimo para encontrarlo jajajajaj

Reply:
i'll implement it in a similar way as you describe above